pkg/storage/tsdb/cacheable_index.go

```diff
@@ -0,0 +1,275 @@
+func (i *CacheableIndex) buildCacheKey(userID string, from, through model.Time, shard *index.ShardAnnotation, matchers ...*labels.Matcher) string {
+	stringifiedShard := fmt.Sprint(shard)
+	keyMembers := []string{
+		userID, from.String(), through.String(), stringifiedShard,
+	}
+	if len(matchers) != 0 {
+		keyMembers = append(keyMembers, stringifiedMatchers(matchers))
+	}
+
+	return strings.Join(keyMembers, sep)
+}
+
+// GetChunkRefs is a cached implementation of GetChunkRefs.
+//
+// It uses as a key for the cache all parameters (userID, from, through, shard, matchers) separated by a colon (:).
+func (i *CacheableIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, res []ChunkRef, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+	fallbackFn := func(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matcher ...*labels.Matcher) ([][]byte, error) {
+		series, err := i.Index.GetChunkRefs(ctx, userID, from, through, res, shard, matcher...)
+		if err != nil {
+			return nil, errors.Wrap(err, "call to GetChunkRefs")
+		}
+
+		var encodeBuf bytes.Buffer
+		enc := gob.NewEncoder(&encodeBuf)
+		if err := enc.Encode(series); err != nil {
+			return nil, errors.Wrap(err, "GetChunkRefs encoding")
+		}
+
+		return [][]byte{encodeBuf.Bytes()}, nil
+	}
+
+	results, err := i.CacheableOp(ctx, "GetChunkRefs", i.buildCacheKey, fallbackFn, userID, from, through, shard, matchers...)
+	if err != nil {
+		return nil, errors.Wrap(err, "cacheable index call to cacheable op GetChunkRef")
+	}
+
+	var values []ChunkRef
+	for _, val := range results {
+		decoderBuf := bytes.NewBuffer(val)
+		dec := gob.NewDecoder(decoderBuf)
+		if err := dec.Decode(&values); err != nil {
+			return nil, errors.Wrap(err, "cacheable index GetChunkRef decoding")
+		}
+
+		res = append(res, values...)
+	}
+
+	return res, nil
+}
+
+// Series is a cached implementation of Series.
+//
+// It uses as a key for the cache all parameters (userID, from, through, shard, matchers) separated by a colon (:).
+func (i *CacheableIndex) Series(ctx context.Context, userID string, from, through model.Time, res []Series, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]Series, error) {
+	fallbackFn := func(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matcher ...*labels.Matcher) ([][]byte, error) {
+		series, err := i.Index.Series(ctx, userID, from, through, res, shard, matcher...)
+		if err != nil {
+			return nil, errors.Wrap(err, "call to Series")
+		}
+
+		var encodeBuf bytes.Buffer
+		enc := gob.NewEncoder(&encodeBuf)
+		if err := enc.Encode(series); err != nil {
+			return nil, errors.Wrap(err, "Series encoding")
+
+		}
+
+		return [][]byte{encodeBuf.Bytes()}, nil
+	}
+
+	results, err := i.CacheableOp(ctx, "Series", i.buildCacheKey, fallbackFn, userID, from, through, shard, matchers...)
```

Comment:
I see we're _always_ turning our results into `[][]byte`, then deserializing them again even when there was a cache miss. I'd like to avoid this considering this code is called on every query.

Reply:
Jordan pushed a commit fixing this, WDYT?